### PR TITLE
Refactor all linear attention models to latest best standards for convolution

### PR DESCRIPTION
**Pull request by Cyrilvallez**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47452)

# What does this PR do?

As per the title. Following the work in https://github.com/huggingface/transformers/pull/47347 for Inkling and being able to rollback the linear cache, I simplified the linear attention conv implementation quite a lot. This brings a full refactor for all other linear attention models (mambas and more) to follow the same approach.
This mostly brings the following:
- add much more inheritance with modular for ease of maintenance in the future
- much much more readable: simplified a lot of dead codepaths
- single implementation everywhere
- the models will be able to restart from previous conv states without loss of states (cache continuation, prefix caching etc)
- the models will be able to rollback their conv states if needed (speculative decoding, e.g. mtp etc)

It's insane the amount of weird dead code paths that the mambas had, for example stuff such as

```python
d_mlp = (
    projected_states.shape[-1]
    - 2 * self.intermediate_size
    - 2 * self.n_groups * self.ssm_state_size
    - self.num_heads
) // 2
z0, x0, gate, hidden_states_B_C, dt = projected_states.split(
        [d_mlp, d_mlp, self.intermediate_size,  self.conv_dim, self.num_heads], dim=-1
)
if d_mlp > 0:
     ...
```

where `d_mlp` will necessarily be `0` from the config, and then we split with the first 2 dims being 0 (so empty tensors), and try to do stuff in branches that cannot be taken... 

Also, all the if/else branches for different cases of cache update were way too complicated and not correct in general, but those simplifications come from the work done in Inkling to make it general and easy.

They also had some config fields that were never used... I removed those, as they are useless (no BC issues since never used).

### CODE REVIEW

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 1]
```diff
@@ -203,15 +244,7 @@ def __init__(self, config: Mamba2Config, layer_idx: int, initialize_mixer_weight
...
-        is_fast_path_available = all(
-            (
-                selective_state_update,
-                mamba_chunk_scan_combined,
-                mamba_split_conv1d_scan_combined,
-                causal_conv1d_fn,
-                causal_conv1d_update,
-            )
-        )
+        is_fast_path_available = all((causal_conv1d, mamba_ssm))
```
Comment (vasqu) [inline src/transformers/models/mamba2/modeling_mamba2.py]: Not sure if I agree with changing this up. In certain cases this saved our ass and revealed bad import paths when we checked that each function actually exists

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 2]
```diff
@@ -238,56 +271,107 @@ def init_mamba2_weights(self):
...
-            # 2. Convolution sequence transformation
-            hidden_states_B_C = causal_conv1d_update(
-                hidden_states_B_C,
-                conv_state,
+            hidden_states = causal_conv1d_fn(
+                hidden_states,
                 self.conv1d.weight.squeeze(1),
                 self.conv1d.bias,
                 self.activation,
+                seq_idx=kwargs.get("seq_idx"),
```
Comment (vasqu) [inline src/transformers/models/mamba2/modeling_mamba2.py:312]: Yea I'd love to just pass **kwargs but the og function would fail, right? Maybe we should revisit the fallback with original so we can also remove all the lazy load dependencies?

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 3]
```diff
@@ -238,56 +271,107 @@ def init_mamba2_weights(self):
...
             )
 
-            hidden_states, B, C = torch.split(
-                hidden_states_B_C,
-                [self.intermediate_size, groups_time_state_size, groups_time_state_size],
-                dim=-1,
+            # Drop the additional previous states
+            if cache_params is not None:
+                hidden_states = hidden_states[:, :, -seq_len:]
```
Comment (vasqu) [inline src/transformers/models/mamba2/modeling_mamba2.py:317]:
```suggestion
            hidden_states = hidden_states[:, :, -seq_len:]
```
I dont think we need to guard tbh

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 4]
```diff
@@ -238,56 +271,107 @@ def init_mamba2_weights(self):
...
+
+        hidden_states = hidden_states.transpose(1, 2)
+        return hidden_states
+
+    def cuda_kernels_forward(
```
Comment (vasqu) [inline src/transformers/models/mamba2/modeling_mamba2.py:322]: Hmm ok so we still keep all of mamba into one big chunk. Maybe we should start splitting this a bit into different sections similar to gdn --> removing the split between torch and cuda and unify into one forward

I think we could have a similar distinction here:
1. Chunk prefill-based implementation
2. One recurrent step based on cached decoding

The only exception which doesn't follow a proper similar pattern is `mamba_split_conv1d_scan_combined` which fuses multiple paths and can be seen as training shortcut

So the overall structure would be
- Projections
- Cache (conv and recurrent state)
- Training shortcut `mamba_split_conv1d_scan_combined`
- Conv (identical to qwen 3 next basically)
  - Update
  - Fn 
- Mamba core op
  - Cache and one seq len --> simple one step == `selective_state_update` or single step torch mamba
  - No cache or greater seq len --> `mamba_chunk_scan_combined` or ssd chunk torch mamba

That way we unify the way we go about this with qwen gdn and we have clearer implementations. If we shift it into the functions we have an even easier time in the future with the kernels approach we want

File: src/transformers/models/qwen3_next/modular_qwen3_next.py [hunk 5]
```diff
@@ -454,19 +454,9 @@ def forward(
...
-        # (single-token decode and chunk-tokens continuation) share the state read here; they only
-        # diverge in how the conv input is assembled and which kernel consumes the states below,
-        # which we gate locally on `seq_len`.
+        seq_len = hidden_states.shape[1]
         use_precomputed_states = cache_params is not None and cache_params.has_previous_state(self.layer_idx)
 
-        # getting projected states from cache if it exists
-        if use_precomputed_states:
-            conv_state = cache_params.layers[self.layer_idx].conv_states[0]
-            recurrent_state = cache_params.layers[self.layer_idx].recurrent_states[0]
```
Comment (vasqu) [inline src/transformers/models/qwen3_next/modular_qwen3_next.py:468]: Would revert this, feels more clean to me to have this on top to extract metadata

File: src/transformers/models/mamba/modeling_mamba.py
Comment (vasqu) [inline src/transformers/models/mamba/modeling_mamba.py:1]: This is where it gets messy: Mamba1 at least in the torch version only has the recurrent single step version, there is no chunk like implementation

But ideally we should do the same structure as mamba2 and gdn as well then. If I have some time I could investigate a chunked implementation

**Review by vasqu — COMMENTED** [review]
Just wanted to leave some initial comments because this has very design based questions. Imo, we should aim to unify the forward in mamba2 with the way we do it in qwen3 next

We should not need the explicit split between cuda and torch and rather go step by step. I left one bigger comment on mamba2 which hopefully explains this better

File: src/transformers/models/mamba/modeling_mamba.py [hunk 6]
```diff
@@ -193,149 +278,117 @@ def cuda_kernels_forward(
...
+        discrete_time_step = self.dt_proj(time_step)
+        discrete_time_step = nn.functional.softplus(discrete_time_step).transpose(1, 2)
 
         # 3.b. Discretization: B and C to [batch, seq_len, intermediate_size, ssm_state_size] (SRAM)
-        A = -torch.exp(self.A_log.float())                                              # [intermediate_size, ssm_state_size]
-        discrete_A = torch.exp(A[None, :, None, :] * discrete_time_step[:, :, :, None]) # [batch, intermediate_size, seq_len, ssm_state_size]
-        discrete_B = discrete_time_step[:, :, :, None] * B[:, None, :, :].float()       # [batch, intermediate_size, seq_len, ssm_state_size]
+        A = -torch.exp(self.A_log.float())
+        discrete_A = torch.exp(A[None, :, None, :] * discrete_time_step[:, :, :, None])
+        discrete_B = discrete_time_step[:, :, :, None] * B[:, None, :, :].float()
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:379]: my shape commmmmments

File: src/transformers/models/mamba/modeling_mamba.py [hunk 7]
```diff
@@ -193,149 +278,117 @@ def cuda_kernels_forward(
...
-                cache_params.update_conv_state(conv_state, self.layer_idx)
-                hidden_states = self.act(self.conv1d(hidden_states)[..., :seq_len])     # [batch, intermediate_size, seq_len]
-            else:
-                conv_state = cache_params.update_conv_state(hidden_states, self.layer_idx)[..., -self.conv_kernel_size:]
-                conv_state = conv_state.to(self.conv1d.weight.device)
-                hidden_states = torch.sum(conv_state * self.conv1d.weight[:, 0, :], dim=-1)
-                if self.use_conv_bias:
-                    hidden_states += self.conv1d.bias
-                hidden_states = self.act(hidden_states).to(dtype).unsqueeze(-1)         # [batch, intermediate_size, 1] : decoding
-        else:
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:308]: noice that was really long long due

File: src/transformers/models/mamba/modeling_mamba.py [hunk 8]
```diff
@@ -167,17 +209,60 @@ def warn_slow_implementation(self):
                     " install the kernels library using `pip install kernels` or https://github.com/Dao-AILab/causal-conv1d for causal-conv1d. For the mamba.py backend, follow https://github.com/alxndrTL/mamba.py."
                 )
 
+    def convolution(
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py]: Given that we now have the dynamic weight loader, and that this functions pretty much holds "weights". I think it would make sense to: 
- create a `CausalConvFn` module
- uniformize with https://github.com/huggingface/transformers/blob/openai-privacy-filter-mps-metal-flash/src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py#L345-L405 

recent audio models also use causal conv1d functions. Some stuff are different since we hold the mamba parameters, but some are similar

File: src/transformers/models/nemotron_h/modeling_nemotron_h.py [hunk 9]
```diff
@@ -176,106 +210,157 @@ def __init__(self, config: NemotronHConfig, layer_idx: int | None = None):
...
+        mamba_ssm = lazy_load_kernel("mamba-ssm")
+        selective_state_update = resolve_internal_import(
+            mamba_ssm, chained_path="ops.triton.selective_state_update.selective_state_update"
+        )
+        mamba_chunk_scan_combined = resolve_internal_import(
+            mamba_ssm, chained_path="ops.triton.ssd_combined.mamba_chunk_scan_combined"
+        )
+        mamba_split_conv1d_scan_combined = resolve_internal_import(
+            mamba_ssm, chained_path="ops.triton.ssd_combined.mamba_split_conv1d_scan_combined"
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron_h/modeling_nemotron_h.py:243]: can 't we solve this with a function decorator like https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L145-L145

File: src/transformers/models/nemotron_h/modeling_nemotron_h.py [hunk 10]
```diff
@@ -176,106 +210,157 @@ def __init__(self, config: NemotronHConfig, layer_idx: int | None = None):
...
         self.layer_type = config.layer_types[layer_idx]
+        self.use_mem_eff_path = True
 
-    def cuda_kernels_forward(
+    @torch.no_grad()
+    def init_nemotron_h_mamba2_weights(self):
+        A = torch.arange(1, self.num_heads + 1, device=self.A_log.device, dtype=torch.float32)
+        init.copy_(self.A_log, torch.log(A))
+        init.ones_(self.D)
+        init.ones_(self.dt_bias)
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron_h/modeling_nemotron_h.py:267]: lol this is another PR where we put this in all module init 👁️ 👁️

File: src/transformers/models/nemotron_h/modeling_nemotron_h.py [hunk 11]
```diff
@@ -297,209 +382,125 @@ def cuda_kernels_forward(
...
             # for batched generation
-            dt = dt[:, None, ...] if dt.ndim == 2 else dt[:, 0, :][:, None, ...]
+            dt = dt[:, 0, :][:, None, ...]
             dt = dt.transpose(1, 2).expand(batch_size, dt.shape[-1], self.head_dim)
-            # [num_heads] -> [num_heads, head_dim]
             dt_bias = self.dt_bias[..., None].expand(self.dt_bias.shape[0], self.head_dim)
 
             dt = torch.nn.functional.softplus(dt + dt_bias.to(dt.dtype))
-            dt = torch.clamp(dt, self.time_step_min) #, self.time_step_max)
+            dt = torch.clamp(dt, self.time_step_limit[0], self.time_step_limit[1])
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron_h/modeling_nemotron_h.py:469]: is this a bug fix?

Comment (ArthurZucker) [reply src/transformers/models/nemotron_h/modeling_nemotron_h.py:469]: s it was previously commented out, would be a bit weird

**Review by ArthurZucker — COMMENTED** [review]
Mostly pushing to create the convolution module will rurther help with modular imo

File: src/transformers/models/nemotron_h/modeling_nemotron_h.py (hunk 11 above)
Comment (Cyrilvallez) [reply src/transformers/models/nemotron_h/modeling_nemotron_h.py:469]: Yep, pretty sure the comment was not intentional looking at all other implems

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 1 above)
Comment (Cyrilvallez) [reply src/transformers/models/mamba2/modeling_mamba2.py]: Oh really? I assumed this would be equivalent, and since the causal_conv1d functions would be always defined now (same name under torch or kernel) it was easier like that, but can switch back

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 2 above)
Comment (Cyrilvallez) [reply src/transformers/models/mamba2/modeling_mamba2.py:312]: Yep definitely, but for another PR IMO, this one focuses mostly on the conv part/simplification of the overall forwards

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 3 above)
Comment (Cyrilvallez) [reply src/transformers/models/mamba2/modeling_mamba2.py:317]: Yep not strictly needed, but I like to explicitly show that it's only relevant if we added the previous state right before the conv if we had a cache

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 4 above)
Comment (Cyrilvallez) [reply src/transformers/models/mamba2/modeling_mamba2.py:322]: Definitely agree that we should simplify the recurrent state update as well, and remove the awkward `kernel_forward`/`torch_forward` in a single forward by defining function with the same name as the kernels and using fallback (the same way I did with the 2 functions for the conv).
However, this PR was aimed mostly at the conv part (even though I simplified the recurrent part quite a bit as well), and it's already quite big like this, so would prefer to do this in the future (would be much clearer/simpler to do after this PR is merged as well, since it adds much more inheritance)

File: src/transformers/models/qwen3_next/modular_qwen3_next.py (hunk 5 above)
Comment (Cyrilvallez) [reply src/transformers/models/qwen3_next/modular_qwen3_next.py:468]: Make more sense IMO to extract the recurrent state only where it's needed, i.e. right after the conv (easier to read as you don't have to scroll back up)

File: src/transformers/models/nemotron_h/modeling_nemotron_h.py (hunk 9 above)
Comment (Cyrilvallez) [reply src/transformers/models/nemotron_h/modeling_nemotron_h.py:243]: Yeah it's ugly but out of scope of this PR to make it clean 🥲

File: src/transformers/models/nemotron_h/modeling_nemotron_h.py (hunk 10 above)
Comment (Cyrilvallez) [reply src/transformers/models/nemotron_h/modeling_nemotron_h.py:267]: It's because of the PR for mamba init that did this, but yeah will disappear into something cleaner when I have the time

File: src/transformers/models/mamba/modeling_mamba.py (hunk 6 above)
Comment (Cyrilvallez) [reply src/transformers/models/mamba/modeling_mamba.py:379]: makes it a bit ugly IMO haha but can add back

**Comment by Cyrilvallez** [comment]
Hey @ArthurZucker @vasqu!

Just a few remarks on the scope of this PR: it's supposed to mostly solve the convolution part of the mambas to make it correct in general (which it is not rn for chunked prefill, cache continuation etc etc) as well as make it much more simple, readable and modular (-3800 LoC, + 2400). Even if I simplified the code paths for recurrent update already, it was not really the expected goal (was just too messy haha I had to refactor a bit).

I definitely agree that we should simplify the recurrent state update further as well, and remove the awkward kernel_forward/torch_forward in a single forward by defining function with the same name as the kernels and using fallback (the same way I did with the 2 functions for the conv).

For the refactor of the weird kernel imports, and the recurrent state part, I would rather do it later as this PR is already big and not so easy to follow from diffs only. It would also be much clearer/simpler to do after this PR is merged, since it adds much more inheritance through modular, so much less work will be needed in a second PR.

So TLDR, further standardization of the recurrent update/or conversion to module is IMO a perfect candidate for future PR, without making this one insanely huge and potentially dangerous/breakable

File: src/transformers/models/nemotron_h/modeling_nemotron_h.py (hunk 11 above)
Comment (Cyrilvallez) [reply src/transformers/models/nemotron_h/modeling_nemotron_h.py:469]: Actually it was intentional, fixed it

**Comment by Cyrilvallez** [comment]
run-slow: bamba, falcon_h1, falcon_mamba, granitemoehybrid, jamba, lfm2, lfm2_moe, lfm2_vl, mamba, mamba2, nemotron_h, qwen3_5, qwen3_5_moe, qwen3_next, zamba, zamba2

**Review by vasqu — APPROVED** [review]
As per internal discussion, let's incrementally go from here. This is already simplifying by a lot so let's not make things even more complicated at once, e.g. -4k +3k 👀 

Only nit: let's mark the conv separation under the module as private until we figure things out step by step
